### PR TITLE
Fix grpc transcoding content type not matching response body

### PR DESCRIPTION
**Pull request by nareddyt**
For a trailer-only gRPC streaming response, ensure that the returned content type is `application/json`. This is required because the response body is an empty JSON array.

Risk Level: Low
Testing: Added an integration test to catch this bug.
Docs Changes: None
Release Notes: None

Fixes: #5011
Ref: #8158
Ref: #8009

Please note this fixes issues for responses that are:
- Trailer-only (either an empty body or a gRPC error)
- Streaming (not unary)

Signed-off-by: Teju Nareddy <nareddyt@google.com>

### CODE REVIEW

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc [hunk 1]
```diff
@@ -403,6 +403,11 @@ Http::FilterHeadersStatus JsonTranscoderFilter::encodeHeaders(Http::HeaderMap& h
     // In gRPC wire protocol, headers frame with end_stream is a trailers-only response.
     // The return value from encodeTrailers is ignored since it is always continue.
     encodeTrailers(headers);
+
+    // When there is no body, a empty JSON array is returned by default. Set the content type
+    // correctly.
+    headers.insertContentType().value().setReference(Http::Headers::get().ContentTypeValues.Json);
```
Comment (qiwzhang) [inline source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: I think encodeTrailers() call may have set the content-type in line L596

**Comment by lizan** [comment]
@nareddyt does this PR supersede #8158?

**Comment by nareddyt** [comment]
Yes, @qiwzhang asked me take it over.

We may have misunderstood the original issue, it seems like the error is only for streaming backends. I need to look into this further (and perhaps add a new integration test)

File: source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc (hunk 1 above)
Comment (nareddyt) [reply source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc]: That should be fine because it just sets it to JSON, never GRPC.

**Comment by nareddyt** [comment]
Ready for review :+1:

**Pull request merged**